src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,571 @@
+     * the end of input elements.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultFinisher()}.
+     *
+     * @return a function which transforms the intermediate result to the final
+     *         result(s) which are then passed on to the provided Downstream
+     */
+    default BiConsumer<A, Downstream<? super R>> finisher() {
+        return defaultFinisher();
+    }
+
+    /**
+     * Returns a composed Gatherer which connects the output of this Gatherer
+     * to the input of that Gatherer.
+     *
+     * @param that the other gatherer
+     * @param <RR> The type of output of that Gatherer
+     * @throws NullPointerException if the argument is null
+     * @return returns a composed Gatherer which connects the output of this
+     *         Gatherer as input that Gatherer
+     */
+    default <RR> Gatherer<T, ?, RR> andThen(Gatherer<? super R, ?, ? extends RR> that) {
+        Objects.requireNonNull(that);
+        return Gatherers.Composite.of(this, that);
+    }
+
+    /**
+     * Returns an initializer which is the default initializer of a Gatherer.
+     * The returned initializer identifies that the owner Gatherer is stateless.
+     *
+     * @see Gatherer#initializer()
+     * @return the instance of the default initializer
+     * @param <A> the type of the state of the returned initializer
+     */
+    static <A> Supplier<A> defaultInitializer() {
+        return Gatherers.Value.DEFAULT.initializer();
+    }
+
+    /**
+     * Returns a combiner which is the default combiner of a Gatherer.
+     * The returned combiner identifies that the owning Gatherer must only
+     * be evaluated sequentially.
+     *
+     * @see Gatherer#finisher()
+     * @return the instance of the default combiner
+     * @param <A> the type of the state of the returned combiner
+     */
+    static <A> BinaryOperator<A> defaultCombiner() {
+        return Gatherers.Value.DEFAULT.combiner();
+    }
+
+    /**
+     * Returns a {@code finisher} which is the default finisher of
+     * a {@code Gatherer}.
+     * The returned finisher identifies that the owning Gatherer performs
+     * no additional actions at the end of input.
+     *
+     * @see Gatherer#finisher()
+     * @return the instance of the default finisher
+     * @param <A> the type of the state of the returned finisher
+     * @param <R> the type of the Downstream of the returned finisher
+     */
+    static <A,R> BiConsumer<A, Downstream<? super R>> defaultFinisher() {
+        return Gatherers.Value.DEFAULT.finisher();
+    }
+
+    /**
+     * Returns a new, sequential, and stateless {@code Gatherer} described by
+     * the given {@code integrator}.
+     *
+     * @param integrator the integrator function for the new gatherer
+     * @param <T> the type of input elements for the new gatherer
+     * @param <R> the type of results for the new gatherer
+     * @throws NullPointerException if the argument is null
```

Comment:
I think we can specify that gatherers are null averse in a single place as we do for Stream an the likes. Then we do not have to nag about @throws NPE

Reply:
I think it's still better to specify this for every method. Many developers read the documentation only for the specific method they are going to call, using IDE features like quick documentation.

Reply:
Yeah, I agree with @amaembo, I think it is important to keep the contract close to the caller.

Reply:
I don't mind either way - just wanted to note that the notion of blanket statements for `NullPointerException` is used in several areas (`[java.util.logging](https://docs.oracle.com/en/java/javase/21/docs/api/java.logging/java/util/logging/package-summary.html)`, `[java.net.http](https://docs.oracle.com/en/java/javase/21/docs/api/java.net.http/java/net/http/package-summary.html)` ...)

Reply:
It is different in  `java.util.stream.Stream` and `Collector` and because `Gatherer` is more similar to `Collector` it might in fact be more consistent having the @throws NPE here.